odoo/addons/base/models/ir_model.py

```diff
@@ -1864,66 +1850,75 @@ def _reflect_constraint(self, model, conname, type, definition, module, message=
+                VALUES (%s,
+                        now() AT TIME ZONE 'UTC',
+                        now() AT TIME ZONE 'UTC',
+                        %s, %s,
+                        (SELECT id FROM ir_module_module WHERE name=%s),
+                        (SELECT id FROM ir_model WHERE model=%s),
+                        %s, %s, %s)
+                RETURNING id
+                """, conname, self.env.uid, self.env.uid, module, model._name, type, definition, Json({'en_US': message})
+            ))
+            return self.browse(cons_id)
+        [cons] = rows
         cons_id = cons.pop('id')
         if cons != dict(type=type, definition=definition, message=message):
-            query = """ UPDATE ir_model_constraint
-                        SET write_date=now() AT TIME ZONE 'UTC',
-                            write_uid=%s, type=%s, definition=%s, message=%s
-                        WHERE id=%s"""
-            cr.execute(query, (self.env.uid, type, definition, Json({'en_US': message}), cons_id))
+            self.env.execute_query(SQL(
+                """
+                UPDATE ir_model_constraint
+                SET write_date=now() AT TIME ZONE 'UTC',
+                    write_uid = %s, type = %s, definition = %s, message = %s
+                WHERE id = %s""",
+                self.env.uid, type, definition, Json({'en_US': message}), cons_id
+            ))
             return self.browse(cons_id)
+        return None
 
     def _reflect_constraints(self, model_names):
-        """ Reflect the SQL constraints of the given models. """
+        """ Reflect the table objects of the given models. """
         for model_name in model_names:
             self._reflect_model(self.env[model_name])
 
     def _reflect_model(self, model):
-        """ Reflect the _sql_constraints of the given model. """
-        def cons_text(txt):
-            return txt.lower().replace(', ',',').replace(' (','(')
-
-        # map each constraint on the name of the module where it is defined
+        """ Reflect the _table_objects of the given model. """
+        # map each object on the name of the module where it is defined
         constraint_module = {
-            constraint[0]: cls._module
+            key: cls._module
             for cls in reversed(self.env.registry[model._name].mro())
             if models.is_definition_class(cls)
-            for constraint in getattr(cls, '_local_sql_constraints', ())
+            for key in cls._local_table_objects
         }
 
         data_list = []
-        for (key, definition, message) in model._sql_constraints:
-            conname = '%s_%s' % (model._table, key)
-            module = constraint_module.get(key)
-            record = self._reflect_constraint(model, conname, 'u', cons_text(definition), module, message)
+        for cons in model._table_objects.values():
+            conname = cons.name(model)
+            module = constraint_module.get(cons.key)
```

Comment:
I have added a `_table_object_definitions` which replaces `_local_table_objects` (as a list, it could be a dict if we wanted to deduplicate names at that point).
I still need to keep `_table_objects` which is similar to what is `_fields`.